### PR TITLE
Add support for PancakeSwap V3

### PR DESCRIPTION
**Pull request by gnarlycow**
This change adds contracts `PCSV3Automan` and `PCSV3OptimalSwapRouter`.  The vast majority of Automan and OptimalSwapRouter code has been moved to base classes that are inherited by UniV3 and PCSV3 implementations.

A deployment script for PCSV3 automan and router is not part of this change.

### CODE REVIEW

File: src/base/SwapRouter.sol [hunk 1]
```diff
@@ -22,24 +29,10 @@ abstract contract SwapRouter is UniV3Immutables, Payments, IUniswapV3SwapCallbac
...
-            assembly {
-                token1 := calldataload(add(data.offset, 0x20))
-            }
-            pay(token1, address(this), pool, uint256(amount1Delta));
-        }
-    }
+    /// @notice Deterministically computes the pool address given the pool key
+    /// @param poolKey The pool key
+    /// @return pool The contract address of the pool
+    function computeAddressSorted(PoolKey memory poolKey) internal view virtual returns (address pool) {}
```
Comment (shuhuiluo) [inline src/base/SwapRouter.sol:35]:
```suggestion
    function computeAddressSorted(PoolKey memory poolKey) internal view virtual returns (address pool);
```

File: src/UniV3Automan.sol [hunk 2]
```diff
@@ -1,827 +1,18 @@
...
-    /// @notice The address list that can perform automation
-    mapping(address => bool) public isController;
-    /// @notice The list of whitelisted routers
-    mapping(address => bool) public isWhiteListedSwapRouter;
+import {PoolAddress} from "@aperture_finance/uni-v3-lib/src/PoolAddress.sol";
+import "./base/Automan.sol";
+import {UniswapV3Callback} from "./base/Callback.sol";
+import {UniV3Immutables} from "./base/Immutables.sol";
 
+contract UniV3Automan is Automan, UniswapV3Callback {
```
Comment (shuhuiluo) [inline src/UniV3Automan.sol:9]:
```suggestion
import "./base/Automan.sol";
import {UniV3Immutables} from "./base/Immutables.sol";
import {UniV3SwapRouter} from "./base/SwapRouter.sol";

contract UniV3Automan is Automan, UniV3SwapRouter {
```

File: src/UniV3Automan.sol [hunk 3]
```diff
@@ -1,827 +1,18 @@
...
-        bytes32 s
-    ) external returns (uint256 newTokenId, uint128 liquidity, uint256 amount0, uint256 amount1) {
-        checkFeeSanity(feePips);
-        checkAuthorizedForToken(tokenId);
-        selfPermitIfNecessary(tokenId, permitDeadline, v, r, s);
-        (newTokenId, liquidity, amount0, amount1) = _rebalance(params, tokenId, feePips, swapData);
-        emit Rebalance(newTokenId);
+    function computeAddressSorted(PoolKey memory poolKey) internal view override returns (address pool) {
+        pool = PoolAddress.computeAddressSorted(factory, poolKey);
     }
```
Comment (shuhuiluo) [inline src/UniV3Automan.sol:17]:
```suggestion
```

File: src/UniV3OptimalSwapRouter.sol [hunk 4]
```diff
@@ -0,0 +1,19 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.18;
+
+import {INonfungiblePositionManager} from "@aperture_finance/uni-v3-lib/src/interfaces/INonfungiblePositionManager.sol";
+import {PoolAddress, PoolKey} from "@aperture_finance/uni-v3-lib/src/PoolAddress.sol";
+import {OptimalSwapRouter, UniswapV3Callback} from "./base/OptimalSwapRouter.sol";
```
Comment (shuhuiluo) [inline src/UniV3OptimalSwapRouter.sol:6]:
```suggestion
import {OptimalSwapRouter, SwapRouter, UniV3SwapRouter} from "./base/OptimalSwapRouter.sol";
```

File: src/UniV3OptimalSwapRouter.sol [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+/// @title Optimal Swap Router
+/// @author Aperture Finance
+/// @dev This router swaps through an aggregator to get to approximately the optimal ratio to add liquidity in a UniV3
+/// pool, then swaps the tokens to the optimal ratio to add liquidity in the same pool.
+contract UniV3OptimalSwapRouter is UniswapV3Callback, OptimalSwapRouter {
+    constructor(INonfungiblePositionManager npm) payable UniV3Immutables(npm) {}
+
+    function computeAddressSorted(PoolKey memory poolKey) internal view override returns (address pool) {
+        pool = PoolAddress.computeAddressSorted(factory, poolKey);
+    }
```
Comment (shuhuiluo) [inline src/UniV3OptimalSwapRouter.sol:18]:
```suggestion
contract UniV3OptimalSwapRouter is SwapRouter, OptimalSwapRouter, UniV3SwapRouter {
    constructor(INonfungiblePositionManager npm) payable UniV3Immutables(npm) {}
```

**Review by shuhuiluo — CHANGES_REQUESTED** [review]

File: src/base/SwapRouter.sol (hunk 1 above)
Comment (gnarlycow) [reply src/base/SwapRouter.sol]: Thank you, Shuhui! I was trying to achieve this but was tricked by VS code into thinking I have to provide a dummy implementation.

**Review by shuhuiluo — APPROVED** [review]

**Review by gaohan137 — APPROVED** [review]

**Pull request merged**